### PR TITLE
feat: Enable placement of liquids from containers

### LINKED ISSUE #10: Better support for filling a bucket already partially full and/or emptying one
May belong more in JS or TTA, although it would be nice to have such utility here.

Can "top off" a portable herbalism station in TTA now and end up with a partially filled bucket. But then can't fill said bucket from a water source :-)

Probably that should be doable if you have a partially filled bucket _or_ there should be an easy way to dump the contents of a bucket. Maybe we can standardize a click system for that. Left click dumps, right click fills? Or right click on anything other than a water source dumps/fills the target if appropriate?

The portable herbalism station has a slot in it for water while the sifting table can be filled by right clicking it while holding a full bucket.

Again pinging @Josharias and @xrtariq2594

### PR DESCRIPTION
**Pull request by 4Denthusiast**
Re-bind picking up liquids with containers to attack (mouse 1), enable placement with activate (mouse 2, formerly pick up liquids), and fix a multiplayer bug relating to creating container images. Fluids registered without a corresponding liquid block (like steam) just can't be placed (except, using the Machines module, in tanks). Liquids can be placed from containers on solid blocks as usual as well as on other liquids.

The AttackEvent, which would usually be activated by mouse button 1, isn't suitable in itself because they're only triggered when there's a target in range to attack, and liquids don't count, so the earlier lower-level AttackRequest event is used instead. This means that I had to re-implement some things from the character system in the FluidAuthoritySystem instead (specifically, part of `fillFluidContainerItem` is based on `CharacterSystem.onAttackRequest`), but with liquids added to the list of targetable things, and I'm not entirely sure I did this correctly with respect to what's meant to happen on the client vs the server. It seems to work though.

In order to test that it was working correctly in multiplayer, I had to also fix an unrelated multiplayer bug. That's the changes to `FluidContainerAssetResolver`.

Machines needs a slight update to avoid placing liquids at the same time as putting fluids in tanks in some cases.

That issue seems to cover a few different things, but I think this resolves #10.

### CODE REVIEW

**Comment by jdrueckert** [comment]
@4Denthusiast needs conflict resolution

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:121]: I don't understand, why we're checking for a raytrace hit here :thinking: can you elaborate a bit?

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:142]: missing line and javadoc

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:143]: same as above

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: why remove this without replacement?

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: For my taste that's a few too many nesting levels... can you flatten that a bit or pull some of it out into an extra function?

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: missing javadoc

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (4Denthusiast) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:121]: When the player clicks with an item, the LocalPlayer casts a ray in the direction the player is facing to determine what this action hits (e.g. where to place it if it's a right click with a block item), but that ray doesn't collide with liquids. In this case, it should be possible to place liquids on other liquids and must be possible to target liquids for removal, therefore the FluidAuthoritySystem does its own raytracing that includes liquids to determine what block to affect.

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (4Denthusiast) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: I removed it because it's no longer applicable, and didn't replace it because the argument names and types seem pretty self-explanatory already.

Comment (pollend) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: just add the params to the javadocs and basically just re-stating the parameter name should be sufficient.

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (pollend) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: I think these nested levels could be replaced with a condition then an empty return.

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (pollend) [inline src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]:
```suggestion
                getLiquidInReach(gazeLocation.getWorldPosition(new Vector3f()), gazeLocation.getWorldDirection(new Vector3f()), character, characterComponent.interactionRange).ifPresent(pos -> {
```

**Review by pollend — CHANGES_REQUESTED** [review]
I have a couple suggestions but I haven't tested this in omega.

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (jdrueckert) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:121]: Interesting, I thought I saw it done differently in other places :thinking: Something about viewing target of the player or something, don't recall exactly unfortunately... Maybe the raytracing was more hidden in those cases through some helper functions from the engine or iota modules :thinking:

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (4Denthusiast) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java]: I don't see the point in that, but fine, I'll do it.

File: src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java
Comment (4Denthusiast) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:121]: There is another event fired by the engine that includes the information about hit location, that's just not usable here because of the issue with it not including liquids.

Comment (jdrueckert) [reply src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java:121]: Ohh okay, understood now :+1:

**Comment by 4Denthusiast** [comment]
I've made all the requested changes I think, but it currently isn't working because of a bug in the engine that was introduced in https://github.com/MovingBlocks/Terasology/pull/4139: the collider for liquids is never set up, so raytracing to find the liquid that the player is pointing at doesn't work. This can still be tested with the engine in its current state by using DebugLiquid, which is a liquid (and can therefore be picked up in containers) but also not penetrable (so it's targetable even with the engine bug). The engine will need to be fixed before this can be merged though.

**Comment by 4Denthusiast** [comment]
I've tested this with the PR https://github.com/MovingBlocks/Terasology/pull/4221, and that worked, so it should be possible to merge this once that engine PR is merged.

**Comment by 4Denthusiast** [comment]
Now the engine PR has been merged, so it should be possible to test and merge this.

**Review by jdrueckert — APPROVED** [review]

**Review by pollend — APPROVED** [review]

**Pull request merged**
